src/primitive_crypto.cc

```diff
@@ -118,6 +120,206 @@ PRIMITIVE(siphash_get) {
+// secret, and the next 8 bytes are transmitted along with each record.
+// Internally these primitives will add 4 more bytes of block counter, starting
+// at 1, to form a 16 byte IV.  It is vital that the nonce is not reused, and
+// this is normally achieved by counting up the explicit part by one for each
+// record that is encrypted.  In TLS this means that this part of the nonce
+// corresponds to the sequence number of the record.
+PRIMITIVE(gcm_start_message) {
+  ARGS(GcmContext, context, int, length, Blob, authenticated_data, Blob, nonce);
+  if (context->remaining_length_in_current_message() != 0) INVALID_ARGUMENT;
+  context->set_remaining_length_in_current_message(length);
+  if (nonce.length() != GcmContext::NONCE_SIZE) INVALID_ARGUMENT;
+  int mode = context->is_encrypt() ? MBEDTLS_GCM_ENCRYPT : MBEDTLS_GCM_DECRYPT;
+  int result = mbedtls_gcm_starts(
+      context->gcm_context(),
+      mode,
+      nonce.address(),
+      nonce.length(),
+      authenticated_data.address(),  // No additional data.
+      authenticated_data.length());
+  if (result != 0) return tls_error(null, process, result);
+
+  return process->program()->null_object();
+}
+
+/**
+If the result byte array was big enough, returns a Smi to indicate how much
+  data was placed in it.
+If the result byte array was not big enough, returns null.  In this case no
+  data was consumed.
+*/
+PRIMITIVE(gcm_add) {
+  ARGS(GcmContext, context, Blob, data, MutableBlob, result);
+  const uint8* data_address = data.address();
+  int data_length = data.length();
+  if (data_length == 0) return Smi::from(0);
+  int remains = context->remaining_length_in_current_message();
+  if (remains < data_length) OUT_OF_BOUNDS;
+  remains -= data_length;
+
+  // Start by copying into the temporary buffer in the context.
+  const int to_copy = Utils::min(
+      GcmContext::BLOCK_SIZE - context->buffered_bytes(),
+      data_length);
+  memcpy(context->buffered_data() + context->buffered_bytes(), data_address, to_copy);
+  data_address += to_copy;
+  data_length -= to_copy;
+  const int buffered = context->buffered_bytes() + to_copy;
+  if (buffered < GcmContext::BLOCK_SIZE && remains != 0) {
+    // Success.  We copied all the data into the internal buffer, and so the
+    // output byte array is inevitably big enough.
+    // Update the context and the number of bytes of new output.
+    context->set_buffered_bytes(buffered);
+    context->set_remaining_length_in_current_message(remains);
+    return Smi::from(0);
+  }
+
+  // Some data is to be encrypted/decrypted.
+  const int to_process_after_internal_buffer = remains == 0
+      ? data_length
+      : Utils::round_down(data_length, GcmContext::BLOCK_SIZE);
+  if (buffered + to_process_after_internal_buffer > result.length()) {
+    // Output byte array not big enough.  At this point we have not yet
+    // modified the context.  Return null to indicate the problem.
+    return process->program()->null_object();
+  }
```

Comment:
This check can be done before doing any memcpy.

Reply:
I prefer to update data_address and data_length when I do the memcpy, and this check uses the updated version of data_length, so I prefer it this way.